superset-frontend/src/explore/reducers/exploreReducer.js

```diff
@@ -121,17 +121,35 @@ export default function exploreReducer(state = {}, action) {
       });
       const hasErrors = errors && errors.length > 0;
 
+      const controls =
+        action.controlName === 'viz_type' &&
+        action.value !== state.controls.viz_type.value
+          ? {
+              ...getControlsState(
+                state,
+                getFormDataFromControls({
+                  ...state.controls,
+                  viz_type: control,
+                }),
+              ),
+              [action.controlName]: {
+                ...control,
+                validationErrors: errors,
+              },
+            }
+          : {
+              ...state.controls,
+              [action.controlName]: {
+                ...control,
+                validationErrors: errors,
+              },
+            };
```

Comment:
These lines are a little difficult to grok, creating a new variable might make it easier to read:

```js
      const currentControlsState =
        action.controlName === 'viz_type' &&
        action.value !== state.controls.viz_type.value
          ? // rebuild the full control state if switching viz type
            getControlsState(
              state,
              getFormDataFromControls({
                ...state.controls,
                viz_type: control,
              }),
            )
          : state.controls;

      return {
        ...state,
        form_data: new_form_data,
        triggerRender: control.renderTrigger && !hasErrors,
        controls: {
          ...currentControlsState,
          [action.controlName]: {
            ...control,
            validationErrors: errors,
          },
        },
      };
```

Reply:
Good point! Done